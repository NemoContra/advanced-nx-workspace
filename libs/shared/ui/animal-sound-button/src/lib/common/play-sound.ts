import { Observable } from 'rxjs';

export const playSound = (src: string): Observable<void> => {
  return new Observable(() => {
    const audio = new Audio(src);
    audio.play();
    return () => audio.pause();
  });
};
