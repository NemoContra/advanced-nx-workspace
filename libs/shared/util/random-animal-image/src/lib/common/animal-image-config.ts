import { InjectionToken } from '@angular/core';

export interface AnimalImageConfig {
  cats: boolean;
  dogs: boolean;
}

export const defaultAnimalImageConfig: AnimalImageConfig = {
  cats: true,
  dogs: true,
};

export const ANIMAL_IMAGE_CONFIG = new InjectionToken<AnimalImageConfig>(
  'ANIMAL_IMAGE_CONFIG'
);
