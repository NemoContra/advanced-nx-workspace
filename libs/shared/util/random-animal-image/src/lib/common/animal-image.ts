export interface AnimalImage {
  src: string;
  alt: string;
}

export interface CatImage {
  file: string;
}

export interface DogImage {
  message: string;
  status: 'success' | 'failure';
}
