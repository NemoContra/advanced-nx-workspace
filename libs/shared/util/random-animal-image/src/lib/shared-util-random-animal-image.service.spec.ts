import {
  ANIMAL_IMAGE_CONFIG,
  AnimalImage,
  AnimalImageConfig,
  CatImage,
  DogImage,
  SharedUtilRandomAnimalImageService,
} from '@flight-workspace/shared/util/random-animal-image';
import {
  createHttpFactory,
  HttpMethod,
  SpectatorHttp,
} from '@ngneat/spectator/jest';
import { Observer } from 'rxjs';
import { mocked } from 'ts-jest/utils';
import { getRandomEntry } from './common/random-from-array';
import { TestRequest } from '@angular/common/http/testing';
import { HttpStatusCode } from '@angular/common/http';

jest.mock('./common/random-from-array', () => ({
  getRandomEntry: jest.fn(),
}));

console.log = jest.fn();

describe('SharedUtilRandomAnimalImageService', () => {
  const createService = createHttpFactory({
    service: SharedUtilRandomAnimalImageService,
  });

  const catImage: CatImage = {
    file: 'https://cat.de/test.png',
  };
  const dogImage: DogImage = {
    status: 'success',
    message: 'https://dog.de/test.png',
  };

  describe.each`
    cats     | dogs     | url                                          | body
    ${true}  | ${true}  | ${'https://aws.random.cat/meow'}             | ${catImage}
    ${true}  | ${false} | ${'https://aws.random.cat/meow'}             | ${catImage}
    ${false} | ${true}  | ${'https://dog.ceo/api/breeds/image/random'} | ${dogImage}
  `('With config cats: $cats and dogs: $dogs', ({ cats, dogs, url, body }) => {
    let spectator: SpectatorHttp<SharedUtilRandomAnimalImageService>;

    beforeEach(() => {
      spectator = createService({
        providers: [
          {
            provide: ANIMAL_IMAGE_CONFIG,
            useValue: { cats, dogs } as AnimalImageConfig,
          },
        ],
      });
    });

    afterEach(() => jest.resetAllMocks());

    it('should get the result successfully', () => {
      const spyObserver: Observer<AnimalImage> = {
        next: jest.fn().mockName('next'),
        error: jest.fn().mockName('error'),
        complete: jest.fn().mockName('complete'),
      };

      const randomEntryMock = mocked(getRandomEntry);

      randomEntryMock.mockImplementation((arr) => arr[0]);

      spectator.service.getAnimalImage().subscribe(spyObserver);

      const testRequest: TestRequest = spectator.expectOne(url, HttpMethod.GET);

      testRequest.flush(body);

      expect(randomEntryMock).toHaveBeenCalledTimes(1);
      expect(randomEntryMock).toHaveBeenCalledWith(expect.any(Array));
      expect(spyObserver).toMatchSnapshot('spyObserver');
    });

    it('should throw an error', () => {
      const spyObserver: Observer<AnimalImage> = {
        next: jest.fn().mockName('next'),
        error: jest.fn().mockName('error'),
        complete: jest.fn().mockName('complete'),
      };

      const randomEntryMock = mocked(getRandomEntry);

      randomEntryMock.mockImplementation((arr) => arr[0]);

      spectator.service.getAnimalImage().subscribe(spyObserver);

      const testRequest: TestRequest = spectator.expectOne(url, HttpMethod.GET);

      testRequest.flush(body, {
        status: HttpStatusCode.InternalServerError,
        statusText: 'Internal Server Error',
      });

      expect(randomEntryMock).toHaveBeenCalledTimes(1);
      expect(randomEntryMock).toHaveBeenCalledWith(expect.any(Array));
      expect(spyObserver).toMatchSnapshot('spyObserver');
    });
  });

  test('With config cats: false and dogs: false', () => {
    const spectator: SpectatorHttp<SharedUtilRandomAnimalImageService> =
      createService({
        providers: [
          {
            provide: ANIMAL_IMAGE_CONFIG,
            useValue: { cats: false, dogs: false } as AnimalImageConfig,
          },
        ],
      });

    const spyObserver: Observer<AnimalImage> = {
      next: jest.fn().mockName('next'),
      error: jest.fn().mockName('error'),
      complete: jest.fn().mockName('complete'),
    };

    const randomEntryMock = mocked(getRandomEntry);

    randomEntryMock.mockImplementation((arr) => arr[0]);

    spectator.service.getAnimalImage().subscribe(spyObserver);

    spectator.controller.expectNone('https://aws.random.cat/meow');
    spectator.controller.expectNone('https://dog.ceo/api/breeds/image/random');

    expect(randomEntryMock).toHaveBeenCalledTimes(1);
    expect(randomEntryMock).toHaveBeenCalledWith(expect.any(Array));
    expect(spyObserver).toMatchSnapshot('spyObserver');
  });
});
