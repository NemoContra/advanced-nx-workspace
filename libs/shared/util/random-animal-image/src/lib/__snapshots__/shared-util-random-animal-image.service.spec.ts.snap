// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SharedUtilRandomAnimalImageService With config cats: false and dogs: false: spyObserver 1`] = `
Object {
  "complete": [MockFunction complete],
  "error": [MockFunction error],
  "next": [MockFunction next],
}
`;

exports[`SharedUtilRandomAnimalImageService With config cats: false and dogs: true should get the result successfully: spyObserver 1`] = `
Object {
  "complete": [MockFunction complete] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "error": [MockFunction error],
  "next": [MockFunction next] {
    "calls": Array [
      Array [
        Object {
          "alt": "A dog",
          "src": "https://dog.de/test.png",
        },
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
}
`;

exports[`SharedUtilRandomAnimalImageService With config cats: false and dogs: true should throw an error: spyObserver 1`] = `
Object {
  "complete": [MockFunction complete],
  "error": [MockFunction error] {
    "calls": Array [
      Array [
        HttpErrorResponse {
          "error": Object {
            "message": "https://dog.de/test.png",
            "status": "success",
          },
          "headers": HttpHeaders {
            "headers": Map {},
            "lazyUpdate": null,
            "normalizedNames": Map {},
          },
          "message": "Http failure response for https://dog.ceo/api/breeds/image/random: 500 Internal Server Error",
          "name": "HttpErrorResponse",
          "ok": false,
          "status": 500,
          "statusText": "Internal Server Error",
          "url": "https://dog.ceo/api/breeds/image/random",
        },
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "next": [MockFunction next],
}
`;

exports[`SharedUtilRandomAnimalImageService With config cats: true and dogs: false should get the result successfully: spyObserver 1`] = `
Object {
  "complete": [MockFunction complete] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "error": [MockFunction error],
  "next": [MockFunction next] {
    "calls": Array [
      Array [
        Object {
          "alt": "A cat",
          "src": "https://cat.de/test.png",
        },
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
}
`;

exports[`SharedUtilRandomAnimalImageService With config cats: true and dogs: false should throw an error: spyObserver 1`] = `
Object {
  "complete": [MockFunction complete],
  "error": [MockFunction error] {
    "calls": Array [
      Array [
        HttpErrorResponse {
          "error": Object {
            "file": "https://cat.de/test.png",
          },
          "headers": HttpHeaders {
            "headers": Map {},
            "lazyUpdate": null,
            "normalizedNames": Map {},
          },
          "message": "Http failure response for https://aws.random.cat/meow: 500 Internal Server Error",
          "name": "HttpErrorResponse",
          "ok": false,
          "status": 500,
          "statusText": "Internal Server Error",
          "url": "https://aws.random.cat/meow",
        },
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "next": [MockFunction next],
}
`;

exports[`SharedUtilRandomAnimalImageService With config cats: true and dogs: true should get the result successfully: spyObserver 1`] = `
Object {
  "complete": [MockFunction complete] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "error": [MockFunction error],
  "next": [MockFunction next] {
    "calls": Array [
      Array [
        Object {
          "alt": "A cat",
          "src": "https://cat.de/test.png",
        },
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
}
`;

exports[`SharedUtilRandomAnimalImageService With config cats: true and dogs: true should throw an error: spyObserver 1`] = `
Object {
  "complete": [MockFunction complete],
  "error": [MockFunction error] {
    "calls": Array [
      Array [
        HttpErrorResponse {
          "error": Object {
            "file": "https://cat.de/test.png",
          },
          "headers": HttpHeaders {
            "headers": Map {},
            "lazyUpdate": null,
            "normalizedNames": Map {},
          },
          "message": "Http failure response for https://aws.random.cat/meow: 500 Internal Server Error",
          "name": "HttpErrorResponse",
          "ok": false,
          "status": 500,
          "statusText": "Internal Server Error",
          "url": "https://aws.random.cat/meow",
        },
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "next": [MockFunction next],
}
`;
